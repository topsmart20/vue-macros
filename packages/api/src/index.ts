export * from '@vue-macros/common'

export * from './vue'
export * from './resolve'
export * from './ts'
export * from './utils'
