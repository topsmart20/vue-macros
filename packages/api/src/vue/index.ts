export * from './analyze'
export * from './emits'
export * from './props'
export * from './types'
export * from './utils'
