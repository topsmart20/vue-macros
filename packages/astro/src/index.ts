import { type AstroIntegration, type ViteUserConfig } from 'astro'
import { type Options, resolveOptions } from 'unplugin-vue-macros'
import { type Plugin } from 'vite'
import VueMacros from 'unplugin-vue-macros/vite'

export type VueMacrosOptions = Options

function findPluginAndRemove(
  name: string,
  plugins: ViteUserConfig['plugins']
): Plugin | undefined {
  const idx = plugins!.findIndex(
    (plugin) => plugin && 'name' in plugin && plugin.name === name
  )
  if (idx === -1) return
  const plugin = plugins![idx]
  plugins!.splice(idx, 1)
  return plugin as any
}

export default function (options?: VueMacrosOptions): AstroIntegration {
  return {
    name: '@vue-macros/astro',
    hooks: {
      'astro:config:setup': ({ config }) => {
        const resolvedOptions = resolveOptions(options ?? {})
        const vue = findPluginAndRemove('vite:vue', config.vite.plugins)
        const vueJsx = findPluginAndRemove('vite:vue-jsx', config.vite.plugins)

        config.vite.plugins?.push(
          VueMacros({
            ...resolvedOptions,
            plugins: {
              vue,
              vueJsx,
            },
          })
        )
      },
    },
  }
}
