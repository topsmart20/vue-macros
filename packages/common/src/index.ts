export * from './vue'
export * from './ast'
export * from './constants'
export * from './unplugin'
export { default as MagicString } from 'magic-string'
