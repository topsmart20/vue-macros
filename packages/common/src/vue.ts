import { parse } from '@vue/compiler-sfc'
import { babelParse } from './ast'
import { getLang } from './lang'
import { REGEX_VUE_SFC } from './constants'
import type { Program } from '@babel/types'
import type { MagicString } from './magic-string'
import type {
  SFCDescriptor,
  SFCParseResult,
  SFCScriptBlock,
} from '@vue/compiler-sfc'

export type _SFCScriptBlock = Omit<
  SFCScriptBlock,
  'scriptAst' | 'scriptSetupAst'
>

export type SFC = Omit<SFCDescriptor, 'script' | 'scriptSetup'> & {
  sfc: SFCParseResult
  script?: _SFCScriptBlock | null
  scriptSetup?: _SFCScriptBlock | null
  lang: string | undefined
  getScriptAst(): Program | undefined
  getSetupAst(): Program | undefined
} & Pick<SFCParseResult, 'errors'>

export function parseSFC(code: string, id: string): SFC {
  const sfc = parse(code, {
    filename: id,
  })
  const { descriptor, errors } = sfc
  const lang = (descriptor.script || descriptor.scriptSetup)?.lang

  return {
    sfc,
    ...descriptor,
    lang,
    errors,
    getSetupAst() {
      if (!descriptor.scriptSetup) return
      return babelParse(descriptor.scriptSetup.content, lang)
    },
    getScriptAst() {
      if (!descriptor.script) return
      return babelParse(descriptor.script.content, lang)
    },
  }
}

export function getFileCodeAndLang(
  code: string,
  id: string
): { code: string; lang: string } {
  if (!REGEX_VUE_SFC.test(id)) return { code, lang: getLang(id) }
  const sfc = parseSFC(code, id)
  const scriptCode = sfc.script?.content ?? ''
  return {
    code: sfc.scriptSetup
      ? `${scriptCode}\n;\n${sfc.scriptSetup.content}`
      : scriptCode,
    lang: sfc.lang ?? 'js',
  }
}

export function addNormalScript({ script, lang }: SFC, s: MagicString) {
  return {
    start() {
      if (script) return script.loc.end.offset

      const attrs = lang ? ` lang="${lang}"` : ''
      s.prependLeft(0, `<script${attrs}>`)
      return 0
    },

    end() {
      if (!script) s.appendRight(0, `\n</script>\n`)
    },
  }
}
