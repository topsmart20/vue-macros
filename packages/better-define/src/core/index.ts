import { MagicString, getTransformResult, parseSFC } from '@vue-macros/common'
import { analyzeSFC } from '@vue-macros/api'
import type { TSEmits, TSProps } from '@vue-macros/api'

export async function transformBetterDefine(
  code: string,
  id: string,
  isProduction?: boolean
) {
  const s = new MagicString(code)
  const sfc = parseSFC(code, id)
  if (!sfc.scriptSetup) return

  const offset = sfc.scriptSetup.loc.start.offset
  const result = await analyzeSFC(s, sfc)
  if (result.props) {
    await processProps(result.props)
  }
  if (result.emits) {
    processEmits(result.emits)
  }

  return getTransformResult(s, id)

  async function processProps(props: TSProps) {
    const runtimeDefs = await props.getRuntimeDefinitions()

    const runtimeDecls = `{\n  ${Object.entries(runtimeDefs)
      .map(([key, { type, required, default: defaultDecl }]) => {
        const escapedKey = JSON.stringify(key)
        let defaultString = ''
        if (defaultDecl) {
          defaultString = defaultDecl('default')
        }
        if (!isProduction) {
          return `${escapedKey}: { type: ${toRuntimeTypeString(
            type
          )}, required: ${required}, ${defaultString} }`
        } else if (type.some((el) => el === 'Boolean' || el === 'Function')) {
          return `${escapedKey}: { type: ${toRuntimeTypeString(
            type
          )}, ${defaultString} }`
        } else {
          // production: checks are useless
          return `${escapedKey}: ${
            defaultString ? `{ ${defaultString} }` : 'null'
          }`
        }
      })
      .join(',\n  ')}\n}`

    let decl = runtimeDecls
    if (props.withDefaultsAst && !props.defaults) {
      // dynamic defaults
      decl = `_BD_mergeDefaults(${decl}, ${s.sliceNode(
        props.withDefaultsAst.arguments[1],
        { offset }
      )})`
      // add helper
      s.prependLeft(
        offset,
        `import { mergeDefaults as _BD_mergeDefaults } from 'vue'`
      )
    }
    decl = `defineProps(${decl})`

    s.overwriteNode(props.withDefaultsAst || props.definePropsAst, decl, {
      offset,
    })
  }

  function processEmits(emits: TSEmits) {
    const runtimeDecls = `[${Object.keys(emits.definitions)
      .map((name) => JSON.stringify(name))
      .join(', ')}]`
    s.overwriteNode(emits.defineEmitsAst, `defineEmits(${runtimeDecls})`, {
      offset,
    })
  }
}

function toRuntimeTypeString(types: string[]) {
  return types.length > 1 ? `[${types.join(', ')}]` : types[0]
}
