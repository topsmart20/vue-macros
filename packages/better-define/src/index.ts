import { createUnplugin } from 'unplugin'
import {
  REGEX_SETUP_SFC,
  REGEX_VUE_SFC,
  REGEX_VUE_SUB,
  createFilter,
  detectVueVersion,
} from '@vue-macros/common'
import { RollupResolve, setResolveTSFileIdImpl } from '@vue-macros/api'
import { transformBetterDefine } from './core'
import type { BaseOptions, MarkRequired } from '@vue-macros/common'
import type { UnpluginContextMeta } from 'unplugin'
import type { PluginContext } from 'rollup'

export interface Options extends BaseOptions {
  isProduction?: boolean
}

export type OptionsResolved = MarkRequired<
  Options,
  'include' | 'version' | 'isProduction'
>

function resolveOptions(
  options: Options,
  framework: UnpluginContextMeta['framework']
): OptionsResolved {
  const version = options.version || detectVueVersion()

  return {
    include: [REGEX_VUE_SFC, REGEX_SETUP_SFC].concat(
      version === 2 && framework === 'webpack' ? REGEX_VUE_SUB : []
    ),
    isProduction: process.env.NODE_ENV === 'production',
    ...options,
    version,
  }
}

const name = 'unplugin-vue-better-define'

export default createUnplugin<Options | undefined, false>(
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework)
    const filter = createFilter(options)

    const { resolve, handleHotUpdate } = RollupResolve()

    return {
      name,
      enforce: 'pre',

      buildStart() {
        if (framework === 'rollup' || framework === 'vite') {
          setResolveTSFileIdImpl(resolve(this as PluginContext))
        }
      },

      transformInclude(id) {
        return filter(id)
      },

      async transform(code, id) {
        try {
          return await transformBetterDefine(code, id, options.isProduction)
        } catch (err: unknown) {
          this.warn(`${name} ${err}`)
          console.warn(err)
        }
      },

      vite: {
        configResolved(config) {
          options.isProduction = config.isProduction
        },

        handleHotUpdate,
      },
    }
  }
)
