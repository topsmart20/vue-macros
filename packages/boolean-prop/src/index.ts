import { type Plugin } from 'vite'
import { type VuePluginApi, getVuePluginApi } from '@vue-macros/common'
import { generatePluginName } from '#macros' assert { type: 'macro' }
import { type Options, transformBooleanProp } from './core/index'

// legacy export
export * from './api'

const name = generatePluginName()

function rollup(options: Options = {}): Plugin {
  let api: VuePluginApi | null | undefined

  return {
    name,
    configResolved(config) {
      try {
        api = getVuePluginApi(config.plugins)
      } catch {}
    },
    buildStart(rollupOpts) {
      if (api === undefined)
        try {
          api = getVuePluginApi(rollupOpts.plugins)
        } catch (error: any) {
          this.warn(error)
          return
        }

      if (!api) return

      api.options.template ||= {}
      api.options.template.compilerOptions ||= {}
      api.options.template.compilerOptions.nodeTransforms ||= []

      api.options.template.compilerOptions.nodeTransforms.push(
        transformBooleanProp(options)
      )
    },
  }
}

export default {
  rollup,
  vite: rollup,
}
