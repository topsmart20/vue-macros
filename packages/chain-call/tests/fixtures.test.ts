import { resolve } from 'node:path'
import { describe } from 'vitest'
import {
  RollupEsbuildPlugin,
  RollupRemoveVueFilePathPlugin,
  RollupVue,
  rollupBuild,
  testFixtures,
} from '@vue-macros/test-utils'
import { transformChainCall } from '../src/core'
import VueChainCall from '../src/rollup'

describe('fixtures', async () => {
  await testFixtures(
    ['tests/fixtures/*.vue', '!tests/fixtures/definePropsRefs.vue'],
    (_, id) =>
      rollupBuild(id, [
        VueChainCall(),
        RollupVue(),
        RollupRemoveVueFilePathPlugin(),
        RollupEsbuildPlugin({
          target: 'esnext',
        }),
      ]),
    {
      cwd: resolve(__dirname, '..'),
      promise: true,
    }
  )
  await testFixtures(
    import.meta.glob('./fixtures/definePropsRefs.vue', {
      eager: true,
      as: 'raw',
    }),
    (_, id, code) => transformChainCall(code, id)?.code
  )
})
