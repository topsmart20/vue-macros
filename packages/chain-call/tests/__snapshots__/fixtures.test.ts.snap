// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fixtures > tests/fixtures/basic.vue 1`] = `
"import { defineComponent } from 'vue';
import { expectTypeOf } from 'expect-type';
import _export_sfc from '[NULL]/plugin-vue/export-helper';

var _sfc_main = /* @__PURE__ */ defineComponent({
  __name: "basic",
  props: {
    foo: { type: String, required: false, default: "111" },
    bar: { type: Number, required: false, default: 111 },
    baz: { type: Boolean, required: false, default: false }
  },
  setup(__props) {
    const props = __props;
    expectTypeOf(props).toEqualTypeOf();
    return () => {
    };
  }
});

var basic = /* @__PURE__ */ _export_sfc(_sfc_main, [__FILE__]);

export { basic as default };
"
`;

exports[`fixtures > tests/fixtures/declare.vue 1`] = `
"import { toDisplayString, defineComponent } from 'vue';
import { expectTypeOf } from 'expect-type';
import _export_sfc from '[NULL]/plugin-vue/export-helper';

var _sfc_main = /* @__PURE__ */ defineComponent({
  __name: "declare",
  props: {
    bar: { type: Boolean, required: false, default: true }
  },
  setup(__props) {
    const props = __props;
    expectTypeOf(props).toEqualTypeOf();
    return (_ctx, _cache) => {
      return toDisplayString(props);
    };
  }
});

var declare = /* @__PURE__ */ _export_sfc(_sfc_main, [__FILE__]);

export { declare as default };
"
`;

exports[`fixtures > tests/fixtures/destruct.vue 1`] = `
"import { openBlock, toDisplayString, createElementBlock, defineComponent } from 'vue';
import _export_sfc from '[NULL]/plugin-vue/export-helper';

var _sfc_main = /* @__PURE__ */ defineComponent({
  __name: "destruct",
  props: {
    foo: { type: String, required: false, default: "111" }
  },
  setup(__props) {
    const { foo } = __props;
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock("div", null, toDisplayString(foo));
    };
  }
});

var destruct = /* @__PURE__ */ _export_sfc(_sfc_main, [__FILE__]);

export { destruct as default };
"
`;

exports[`fixtures > tests/fixtures/empty.vue 1`] = `
"import { defineComponent } from 'vue';
import { expectTypeOf } from 'expect-type';
import _export_sfc from '[NULL]/plugin-vue/export-helper';

var _sfc_main = /* @__PURE__ */ defineComponent({
  __name: "empty",
  setup(__props) {
    const props = __props;
    expectTypeOf().toEqualTypeOf();
    return () => {
    };
  }
});

var empty = /* @__PURE__ */ _export_sfc(_sfc_main, [__FILE__]);

export { empty as default };
"
`;

exports[`fixtures > tests/fixtures/empty-defaults.vue 1`] = `
"import { defineComponent } from 'vue';
import { expectTypeOf } from 'expect-type';
import _export_sfc from '[NULL]/plugin-vue/export-helper';

var _sfc_main = /* @__PURE__ */ defineComponent({
  __name: "empty-defaults",
  props: {
    foo: { type: String, required: false }
  },
  setup(__props) {
    const props = __props;
    expectTypeOf().toEqualTypeOf();
    return () => {
    };
  }
});

var emptyDefaults = /* @__PURE__ */ _export_sfc(_sfc_main, [__FILE__]);

export { emptyDefaults as default };
"
`;

exports[`fixtures > tests/fixtures/tsx.vue 1`] = `
"import { defineComponent } from 'vue';
import _export_sfc from '[NULL]/plugin-vue/export-helper';

var _sfc_main = /* @__PURE__ */ defineComponent({
  __name: "tsx",
  props: {
    foo: { type: String, required: false, default: "111" }
  },
  setup(__props) {
    return () => {
    };
  }
});

var tsx = /* @__PURE__ */ _export_sfc(_sfc_main, [__FILE__]);

export { tsx as default };
"
`;
